llvm/lib/Target/AMDGPU/AMDGPUReflect.cpp

```diff
@@ -0,0 +1,82 @@
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This pass searches for occurences of the AMDGPU_OCLC_REFLECT function, and
+// replaces the calls with some val dependent on the operand of the func. This
+// can be used to reflect across different implementations of functions at
+// compile time based on a compiler flag or some other means. The first use case
+// is to choose a safe or unsafe version of atomic_xor at compile time, which
+// can be chosen at compile time by setting the flag -amdgpu-unsafe-xor=true.
+//
+// This pass is similar to the NVPTX pass NVVMReflect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "AMDGPU.h"
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/InstIterator.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+
+#define AMDGPU_OCLC_REFLECT "__oclc_amdgpu_reflect"
+
+static cl::opt<bool>
+    AMDGPUReflectEnabled("amdgpu-reflect-enable", cl::init(true), cl::Hidden,
+                         cl::desc("AMDGPU reflection, enabled by default"));
+static cl::opt<bool>
+    AMDGPUUnsafeXorEnable("amdgpu-unsafe-atomic-xor", cl::init(false),
+                          cl::Hidden,
+                          cl::desc("AMDGPU reflection, enabled by default"));
+
+PreservedAnalyses AMDGPUReflectPass::run(Function &F,
+                                         FunctionAnalysisManager &AM) {
+  if (!AMDGPUReflectEnabled) {
+    return PreservedAnalyses::all();
+  }
+
+  if (F.getName() == AMDGPU_OCLC_REFLECT) {
+    assert(F.isDeclaration() && "_reflect function should not have a body");
+    return PreservedAnalyses::all();
+  }
+
+  SmallVector<Instruction *, 4> ToRemove;
+
+  for (Instruction &I : instructions(F)) {
+    CallInst *Call = dyn_cast<CallInst>(&I);
+    if (!Call)
+      continue;
+    Function *Callee = Call->getCalledFunction();
+    if (!Callee || Callee->getName() != AMDGPU_OCLC_REFLECT)
+      continue;
+
+    assert(Call->getNumOperands() == 2 &&
+           "Wrong number of operands to __oclc_amdgpu_reflect function");
+
+    const Value *Str = Call->getArgOperand(0);
+    assert(isa<Constant>(Str) &&
+           "Format of __oclc_amdgpu_reflect function not recognized");
+
+    const Value *Operand = cast<Constant>(Str)->getOperand(0);
+    StringRef ReflectArg = cast<ConstantDataSequential>(Operand)->getAsString();
+    ReflectArg = ReflectArg.substr(0, ReflectArg.size() - 1);
+
+    if (ReflectArg == "AMDGPU_OCLC_UNSAFE_XOR") {
+      int ReflectVal = AMDGPUUnsafeXorEnable ? 1 : 0;
+      Call->replaceAllUsesWith(ConstantInt::get(Call->getType(), ReflectVal));
```

Comment:
You're modifying the function while iterating over it. That's not allowed. You'll have to replace *and* erase your call instructions when you loop over `ToRemove` below

Reply:
Thanks change made.